apps/passport-server/src/database/queries/pipelineSemaphoreHistoryDB.ts

```diff
@@ -0,0 +1,155 @@
+import { QueryResultRow } from "pg";
+import { Pool } from "postgres-pool";
+import { sqlQuery } from "../sqlQuery";
+
+export interface IPipelineSemaphoreHistoryDB {
+  getLatestGroupsForPipeline(
+    pipelineId: string
+  ): Promise<PipelineSemaphoreGroupHistory[]>;
+  add(
+    pipelineId: string,
+    groupId: string,
+    rootHash: string,
+    serializedGroup: string
+  ): Promise<void>;
+  getLatestHistoryForGroup(
+    pipelineId: string,
+    groupId: string
+  ): Promise<PipelineSemaphoreGroupHistory | undefined>;
+  getHistoricalGroup(
+    pipelineId: string,
+    groupId: string,
+    rootHash: string
+  ): Promise<PipelineSemaphoreGroupHistory | undefined>;
+  deleteGroup(pipelineId: string, groupId: string): Promise<void>;
+}
+
+export class PipelineSemaphoreHistoryDB implements IPipelineSemaphoreHistoryDB {
+  private db: Pool;
+
+  public constructor(db: Pool) {
+    this.db = db;
+  }
+
+  /**
+   * Gets the latest Semaphore group data for each group on a pipeline.
+   */
+  public async getLatestGroupsForPipeline(
+    pipelineId: string
+  ): Promise<PipelineSemaphoreGroupHistory[]> {
+    const result = await sqlQuery(
+      this.db,
+      `
+      SELECT DISTINCT ON (group_id) id, pipeline_id, group_id, root_hash, serialized_group, time_created
+      FROM generic_issuance_semaphore_history
+      WHERE pipeline_id = $1
+      ORDER BY group_id, id DESC`,
+      [pipelineId]
+    );
+
+    return result.rows.map(rowToGroupHistory);
+  }
+
+  /**
+   * Returns the latest history entry for a specific group.
+   */
+  public async getLatestHistoryForGroup(
+    pipelineId: string,
+    groupId: string
+  ): Promise<PipelineSemaphoreGroupHistory | undefined> {
+    const result = await sqlQuery(
+      this.db,
+      `
+      SELECT DISTINCT ON (group_id) id, pipeline_id, group_id, root_hash, serialized_group, time_created
```

Comment:
It's used for ordering. `ORDER BY group_id, id DESC` means "for entries with the same group ID, order them by the `id` column", and since we only want one entry, we get the one with the highest value for `id`. Admittedly this is quite subtle and should have been better documented.